### ISSUE TITLE
Two times set-cookie

### ISSUE THREAD

**Issue by adamus007p** [PrestaShop/PrestaShop#29868]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

I have noticed that Prestashop send two time cookies:

- set-cookie: domain
- set-cookie: domain again

```
curl -I https://www.domain.com`
`HTTP/1.1 200 OK
Date: Tue, 04 Oct 2022 19:11:40 GMT
Content-Type: text/html; charset=utf-8
Connection: keep-alive
vary: Accept-Encoding
expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
pragma: no-cache`
`set-cookie: domain-31e23a2a78cb8898c597880b00009cxxx; expires=Mon, 24-Oct-2022 19:11:40 GMT; Max-Age=1728000; path=/; domain=www.domain.com; secure; HttpOnly; SameSite=Lax`
`set-cookie: domain-31e23a2a78cb8898c597880b00009c19xxxxx; expires=Mon, 24-Oct-2022 19:11:40 GMT; Max-Age=1728000; path=/; domain=www.domain.com; secure; HttpOnly; SameSite=Lax`
`x-powered-by: centminmod
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
CF-Cache-Status: DYNAMIC
Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=QZ9xhWXbb7Sh%2FdSRIwEcFdxxxxxhVXASa%2FdFGJIf1Ab7wNFpfi"}],"group":"cf-nel","max_age":604800}
NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
Strict-Transport-Security: max-age=31536000; preload
Server: cloudflare
CF-RAY: 755039845ff26987-FRA
```

I use nginx and cloudflare.
I have tested without cloudlare and the result is the same.

- CentOS Version:        7.9
- Centmin Mod:           124.00stable.s53
- Nginx Version:         1.21.6 (031022-221318-centos7-kvm-f6624e7-br-6e975bc)
- PHP-FPM Version:       7.4.32
- Prestashop 1.7.8.8

### Steps to reproduce

`curl -I https://www.domain.com`

![image](https://user-images.githubusercontent.com/23308427/194537583-929b0c01-60e0-4e16-b575-8c408a5d11ff.png)

### PrestaShop version(s) where the bug happened

8.0.0

**Comment by MhiriFaten** [comment]
Hello @adamus007p ,

I think you mean 1.7.8.7 cause we didn't release 1.7.8.8 yet :sweat_smile: 
Could you please provide a screen record :pray: 
Waiting for your feedback.

Thanks!

**Comment by adamus007p** [comment]
Yes I mean that :) I have installed the newest version from github.

I attach the screen.
![screen-setcookies](https://user-images.githubusercontent.com/19956106/194073153-b35ebe58-1d0b-441b-ba8a-05b2c02b9af5.png)

**Comment by hibatallahAouadni** [comment]
Hello @adamus007p 

Thanks for reporting the issue!
Ping @PrestaShop/prestashop-core-developers someone please try to reproduce the issue, it's too technical!

Thanks in advance!

**Comment by adamus007p** [comment]
@hibatallahAouadni

I am reproducing it in Prestashop demo to showing that it is not related to my serwer. I tested it too to see if it is related to Cloudflare. It is not. It is related to Prestashop.

`curl -I https://mundane-desire.demo.prestashop.com/
`
Set-Cookie: two times.

![image](https://user-images.githubusercontent.com/19956106/194705480-3d5b8445-f670-4319-9a04-131b4c7870c8.png)

**Comment by adamus007p** [comment]
What I have noticed too that Prestashop send cookie in the end of connection. What are purposes for that? I was trying to analize the code.

**Comment by atomiix** [comment]
@hibatallahAouadni I confirm I see the set-cookie twice. There's something to investigate here!

**Comment by adamus007p** [comment]
deleted

**Comment by adamus007p** [comment]
@atomiix @hibatallahAouadni I see that this second cookie comes from the module:
Ps_categorytree.php > setLastVisitiedCategory

**Comment by boo-code** [comment]
Still reproduces on 9.2.x - two `Set-Cookie` headers for the same `PrestaShop-<hash>` cookie, with different payloads, on every front-office page:

```
Set-Cookie: PHPSESSID=...
Set-Cookie: PrestaShop-e00ac869...=def50200 7fdfc91c ...
Set-Cookie: PrestaShop-e00ac869...=def50200 956ac258 ...
```

It is not `ps_categorytree`, or at least not only - the home page has no category and shows it too.

Instrumenting `Cookie::write()` to record what it emits and from where gives the mechanism directly. On a plain home-page request:

```
EMIT keys=date_add,id_lang,id_currency
    FrontControllerCore->smartyOutputContent @ FrontController.php:757
    FrontControllerCore->display @ Controller.php:328
    ControllerCore->run @ Dispatcher.php:508

EMIT keys=date_add,id_lang,id_currency,id_guest,id_connections
    CookieCore->__destruct
```

So `FrontController::smartyOutputContent()` deliberately writes the cookie before flushing the page:

```php
protected function smartyOutputContent($content)
{
    $this->context->cookie->write();
```

and then `id_guest` and `id_connections` are added to the cookie **after** that point, so the destructor's `write()` sees changed content and emits a second header. `Cookie::write()` is otherwise well guarded - it returns early unless `_modified` is set and the checksum actually differs - so this is not a spurious repeat, it is genuinely a second state being persisted.

That also explains @adamus007p's observation that a cookie is sent "in the end of connection": the second one is the shutdown write.

Fixing it means deciding where the visitor-tracking keys should be populated relative to the pre-output write - either they are set before `smartyOutputContent()` writes, so one header carries the final state, or the early write is dropped and only the destructor writes (which changes when headers are emitted relative to output, so it needs care with output buffering). Both are ordering decisions rather than a local patch, which fits the `Needs Specs` label, so I have not touched it. Can prepare whichever direction is preferred.
